fms/utils/serialization.py

```diff
@@ -88,48 +107,117 @@ def get_adapted(
+    """
+    model_path = Path(os.path.expanduser(model_path))
+    if (
+        model_path.suffix == ".safetensors"
+        or len(sorted(model_path.glob("*.safetensors"))) > 0
+    ):
+        return "st"
+    if model_path.suffix == ".pth" or len(sorted(model_path.glob("*.pth"))) > 0:
+        return "pt"
+    if model_path.suffix == ".bin" or len(sorted(model_path.glob("*.bin"))) > 0:
+        return "hf"
+    return "pt_all"
+
+
+def get_safetensors_item(key, file: Path, device: torch.device) -> torch.Tensor:
+    from safetensors import safe_open  # type: ignore[import-untyped]
+
+    with torch.no_grad():
+        with safe_open(
+            file, framework="pt", device=str(device)
+        ) as model_weights:  # type: ignore[attr-defined]
+            return model_weights.get_tensor(key)
+
+
+class LazySafetensorsDict(collections.UserDict):
+    def set_lazy_tensor(self, key, file, device):
+        super().__setitem__(key, lambda: get_safetensors_item(key, file, device))
+
+    def __getitem__(self, key):
+        lazy_tensor = super().__getitem__(key)
+        if callable(lazy_tensor):
+            lazy_tensor = lazy_tensor()
+            super().__setitem__(key, lazy_tensor)
+        return lazy_tensor
+
+
 def load_state_dict(
     model_path: Union[str, Path],
+    *,
+    checkpoint_format: Optional[str] = None,
     distributed_strategy: Optional[str] = None,
     checkpoint_sharding: Optional[str] = None,
     initial_device: torch.device = torch.device("cpu"),
     rank: int = 0,
     world_size: int = 1,
-) -> Mapping[str, Any]:
+) -> MutableMapping[str, Any]:
     """
     Validates that the file(s) found at a checkpoint path are compatible with
-    the intended (possibly distributed) use-case, and returns the state dict
-    if needed.
+    the intended (possibly distributed) use-case, and returns a lazy loading
+    state dict if possible (some formats may not support that).
+
+    If model_path is a directory, it'll try to load, in this order, pytorch
+    models (.pth format), HF models (.bin format), and safetensors
+    (.safetensors format), unless checkpoint_format is specified.
 
     Args:
     model_path: the path to find the weights. If not set, return None.
+    checkpoint_format: how the checkpoint files are saved: None, 'pt',
+            'hf', or 'st'. If None, guess based on files.
     distributed_strategy: the kind of possibly-distributed model in which we
             intend to load these weights. E.g. tp, fsdp, None. Used for
             validation.
     checkpoint_sharding: the sharding format of the checkpoint.
             E.g. layer, tp, fsdp.
-    initial_device: where the state dict will be loaded. if meta, return None.
+    initial_device: where the state dict will be loaded if not lazy.
+            If meta, return empty dict.
     """
     if model_path is None or initial_device.type == "meta":
         return {}
-    # TODO: Add support for tp-sharding a non-sharded state dict.
```

Comment:
curious why removing the todo.. change mind? or replaced by below text (for now only TP)

Reply:
This PR adds support for TP sharding a non-TP checkpoint, specifically non-sharded and layer sharded (a la hf), so the TODO is resolved

Reply:
thx!! misread the TODO :)